nemo array-range to loop transformation [closes #843]

rupertford opened the pull request.

rupertford commented:
Ready for first review subject to CI being happy.

rupertford commented:
Taking this back temporarily as there is a dictionary ordering issue for variable names (unrelated to this PR) that the new tests have uncovered.

rupertford commented:
OK, all seems well. Definitely ready for review from @arporter.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+            try:
+                _ = int(lower_bound_info)
+                lower_bound = Literal(lower_bound_info, INTEGER_TYPE)
+            except ValueError:
```
I think this may need a bit more care - e.g. what if a user has made a mistake editing the config file and got "1a" when they meant "1"? Does the configuration-file reader do any validation (i.e. check for valid integer or Fortran variable name)? If not, I think that's a new issue.

rupertford replied:
No, the config-file reader just gets hold of the string and passes it on. I'll make a new issue and reference it.

rupertford replied:
Issue is #1035

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+        # Upper bound
+        if not array_reference.is_upper_bound(array_index):
+            # The range specifies an upper bound so use it
+            upper_bound = node.start
```
Should be `node.stop`? Implies a missing/incomplete test?

rupertford replied:
Well spotted. The test is there, I had just cut-n-pasted the output into the test without checking properly (see `test_apply_fixed_bounds`). Fixed now.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+    :raises IndexError: if the array does not contain a Range node.
+
+    '''
+    idx = len(array.children)-1
```
I think this could be nicer, e.g:

    for child in reversed(array.children):
        if isinstance(child, Range):
            break
    else:
        raise IndexError
    return child.position

rupertford replied:
Yes, much better. I changed your proposed code to `return child.position` instead of `break` which allowed me to remove the else and just `raise IndexError` after the loop.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+            symbol_table.add(loop_variable_symbol)
+
+        # Replace the loop_idx array dimension with the loop variable.
+        for array in assignment.walk(ArrayReference):
```
I think this needs more care. e.g. NEMO contains code like:

    my_array(:,:) = input(:,:) + some_fn(input2(:,:,:))

rupertford replied:
Yes, that was sloppy. I think the transformation is now robust. It should now only making changes where it is valid to do so. I've supported this by only modifying a range if all parents of the array reference that contains this range node are operators, up to the assignment node. The one additional case I've ruled out is array valued operators and a transformation error is now raised if one of these (currently only matmul) is found. The reason for this is that we need to know the dimension of the array returned by the operator to determine how many dimensions (if any) could be transformed to explicit loops.

arporter reviewed (changes requested):
I'm going to pause this review here as it's the weekend and there's a few things to sort out in what I've looked at so far.

rupertford commented on doc/user_guide/nemo.rst:
```diff
@@ -154,6 +155,35 @@ Internal Representation of it::
+fparser2 ast. This means that any changes made by these
+transformations will not be observed in the code that is output by
+psyclone. The changes will only become visible when the PSyIR
+back-ends replace the current fparser2 ast-based backend.
```
Thanks.

rupertford commented on doc/user_guide/nemo.rst:
```diff
@@ -154,6 +155,35 @@ Internal Representation of it::
+.. warning:: the transformation below all operate on the PSyIR not the
+fparser2 ast. This means that any changes made by these
+transformations will not be observed in the code that is output by
+psyclone. The changes will only become visible when the PSyIR
```
Oops.

rupertford commented on doc/user_guide/nemo.rst:
```diff
@@ -154,6 +155,35 @@ Internal Representation of it::
+
+The following transformations are specific to the NEMO API.
+
+.. warning:: the transformation below all operate on the PSyIR not the
```
Typo and formatting now fixed.

rupertford commented on doc/user_guide/nemo.rst:
```diff
@@ -154,6 +155,35 @@ Internal Representation of it::
+
 .. _limitations:
 
 Limitations
```
Good point. I ignored 1 and 6 originally as the new transformations only work with the PSyIR back-ends so I thought any limitations should still stand. However, 1 and 6 should have been removed when we removed the original nemo-specific loop-fuse transformation so I've now done it. I've also added the point that the transformations will not modify the output code at the moment.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+# -----------------------------------------------------------------------------
+# Author R. W. Ford, STFC Daresbury Lab
+
+'''Module providing a transformation from a PSyIR Array Range to a
```
Oops.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+    >>> from psyclone.psyir.nodes import Assignment
+    >>> from psyclone.domain.nemo.transformations import \
+    >>>     NemoAllArrayRange2LoopTrans
+    >>> from psyclone.errors import TransformationError
```
Oops.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+    def apply(self, node, options=None):
+        '''Apply the NemoAllArrayRange2Loop transformation to the specified
+        node. The node must be an assignment. All range nodes
+        in each array within the assignment are replaced with a loop
```
Oops.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+        node. The node must be an assignment. All range nodes
+        in each array within the assignment are replaced with a loop
+        index with the expected name for the particular loop dimension
+        within a NemoLoop iterating over that index. The bounds of the
```
I've rewritten the words here.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dictionary of string:values or None
```
done.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+    @property
+    def name(self):
+        '''
+        :returns: the name of the transformation as a string.
```
OK.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dictionary of string:values or None
```
updated.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+        try:
+            while True:
+                trans.apply(node)
+        except TransformationError:
```
I think that this is what the HPC expert would want. The docstring example shows why - we can simply iterate over all assignments and apply the transformation without needing to worry about whether there is an exception or not. As for correctness the NemoOuterArrayRange2LoopTrans and NemoArrayRange2LoopTrans validate methods should ensure that this is only applied if it is valid to do so.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+                "supplied node argument should be a PSyIR Assignment, but "
+                "found '{0}'.".format(type(node).__name__))
+
+
```
Done.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+
+'''Module providing a transformation from a PSyIR ArrayReference Range
+to a PSyIR NemoLoop. This is useful for capturing the contents of
+array ranges as kernel regions so they can be optimised.
```
I've rewritten the description.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dictionary of string:values or None
```
No need to be rude.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dictionary of string:values or None
```
Now I'm insulted.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_allarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,140 @@
+
+'''Module providing a transformation from a PSyIR Array Range to a
+PSyIR NemoLoop. This is useful for capturing the contents of array
+ranges as kernel regions so they can be optimised.
```
This has been re-written

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within an Assignment node, but found '{0}'."
+                .format(type(array_ref.parent).__name__))
```
It will return None if parent is None which I think is what it should do. Test `test_within_assignment` in `nemo_arrayrange2loop_trans_test.py` checks for this.

arporter replied:
Oh yes, ignore me.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+        # transformation would be invalid).
+        my_index = node.parent.children.index(node)
+        num_siblings = len(node.parent.children)
+        for idx in range(my_index+1, num_siblings):
```
Much nicer, thanks.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+            following_node = node.parent.children[idx]
+            if isinstance(following_node, Range):
+                raise TransformationError(
+                    "Error in NemoArrayRange2LoopTrans transformation. This "
```
Well there is already a transformation by that name but it is applied to the Assignment node. I do take your point though. We did discuss what the lowest level transformation should be and I thought we agreed that it should be one that is applied to a specified index, which is why I've done it this way. However, I don't think we should allow the transformation to be applied if it is not the outermost Range as that could result in things being computed in a different order. So, should I keep this transformation or drop it?

arporter replied:
Ah, I see. I think we should keep it but maybe update the docstrings to make it clear - hence my confusion over "outermost" earlier.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+
+    def __str__(self):
+        return (
+            "Convert the PSyIR assignment for a specified ArrayReference "
```
I'm not sure I agree. That is a constraint. We don't specify all constraints in the str representation of a class.

arporter replied:
OK.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+        self.validate(node)
+
+        array_reference = node.parent
+        array_index = array_reference.children.index(node)
```
Oh yes. Thanks.

rupertford commented on doc/user_guide/nemo.rst:
```diff
@@ -154,6 +155,35 @@ Internal Representation of it::
+
+.. autoclass:: psyclone.domain.nemo.transformations.NemoArrayRange2LoopTrans
+    :noindex:
+    :members: apply, validate
```
I've updated all of the examples in the docstrings.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,275 @@
+    '''Provides a transformation from a PSyIR ArrayReference Range to a
+    PSyIR NemoLoop. For example:
+
+    >>> schedule = ...
```
I've updated the example. It is now applied to `tra_adv.F90`.

rupertford commented:
Ready for next review from @arporter.

arporter commented on doc/user_guide/nemo.rst:
```diff
@@ -1,7 +1,7 @@
 .. BSD 3-Clause License
 ..
-.. Copyright (c) 2018-2019, Science and Technology Facilities Council.
+.. Copyright (c) 2018-2020, Science and Technology Facilities Council.
```
21 now please.

rupertford replied:
git log shows the last update to this file being Dec 2020. Updated to 2021 anyway.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,346 @@
+
+    '''
+    def apply(self, node, options=None):
+        '''Apply the NemoOuterArrayRange2Loop transformation if the supplied
```
Should that say "NemoArrayRange2Loop" ?

rupertford replied:
Oops.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,346 @@
+        node is the outermost Range node specifying an access to an
+        array index within an Array Reference that is on the
+        left-hand-side of an Assignment node. If this is the case then
+        the outermost Range nodes within array references within the
```
Does this need re-writing now? A quick scan of the code doesn't seem to show any need for the supplied range to be outermost?

rupertford replied:
The generated code would be incorrect if the transformation is applied to a Range node that is not outermost. This is checked in the validate() method.

rupertford replied:
As you suggest later, I've tried to make this clearer in the documentation.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,346 @@
+    >>> schedule.view()
+    >>>
+    >>> from psyclone.psyir.nodes import Range
+    >>> from psyclone.domain.nemo.transformations import \
```
Picky (might as well start the year as I mean to go on) but you don't need the ">>>" after the line-continuation symbol here as it makes for strangely-rendered code in the User Guide.

rupertford replied:
I don't think that is picky :-)

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -0,0 +1,346 @@
+        # return arrays as this is not currently supported?
+        for operation in assignment.rhs.walk(Operation):
+            # At the moment the only array valued operator is matmul
+            if operation.operator == BinaryOperation.Operator.MATMUL:
```
Really we should have a list of array-valued operators but it's a small point and can be done later.

rupertford replied:
Yes, that would be better. It is also possible that #685 and #688 would provide an alternative solution as we could determine what the return type is instead of keeping a list.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_outerarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,186 @@
+    >>>
+    >>> from psyclone.psyir.nodes import Assignment
+    >>> from psyclone.domain.nemo.transformations import \
+    >>>     NemoOuterArrayRange2LoopTrans
```
No ">>>" here.

rupertford replied:
Updated, thanks.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_outerarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,186 @@
+
+    def validate(self, node, options=None):
+        '''Perform various checks to ensure that it is valid to apply the
+        NemoArrayRange2LoopTrans transformation to the supplied PSyIR Node.
```
"NemoOuterArrayRange2LoopTrans"

rupertford replied:
Oops.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_outerarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,186 @@
+            to None.
+        :type options: dict of string:values or None
+
+        :raises TransformationError: if the supplied node is an \
```
s/is an/is not an/ I think?

rupertford replied:
Oops.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_outerarrayrange2loop_trans.py:
```diff
@@ -0,0 +1,186 @@
+        array_reference = node.lhs
+        # Has the array reference got a range?
+        try:
+            _ = _get_outer_index(array_reference)
```
Since we're using this method in this module, it can no longer claim to be private and therefore shouldn't have the leading underscore in its name.

rupertford replied:
This is a situation where I don't want the function visible to the user via the API but I do want to use it in more than one class (in different files). There does not seem to be a Python solution for this and I went for using `_` to hide it from the API. However, I'm happy to remove it as requested.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -474,11 +475,73 @@ class NemoLoop(Loop):
+        # Indicate the type of loop
+        loop_type_mapping = Config.get().api_conf("nemo") \
+                                        .get_loop_type_mapping()
+        if variable.name in loop_type_mapping:
```
You could do:

    loop.loop_type = loop_type_mapping.get(variable.name, "unknown")

rupertford replied:
Much nicer thanks. I don't think I've used that form before.

arporter reviewed (changes requested):
Looking good now. I've only done the source files so far and will look at the tests and test files next time.

rupertford commented:
I think I've covered all the outstanding points. if so, ready for next review from @arporter assuming that the CI is happy.

arporter commented on src/psyclone/tests/nemo/test_files/implicit_different_dims.f90:
```diff
@@ -0,0 +1,41 @@
+! -----------------------------------------------------------------------------
+! Author R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+program implicit_many_dims
```
Please could you change the program name to match the name of the file to avoid confusion.

rupertford replied:
Oops.

arporter commented on src/psyclone/tests/nemo/test_files/implicit_single_value.f90:
```diff
@@ -0,0 +1,41 @@
+  implicit none
+  integer :: jpi, jpj, jpk, jpt, ndim
+  real, dimension(jpi,jpj,jpk,jpt,ndim) :: umask, vmask
+
```
Please add a comment explaining the purpose of this file.

rupertford replied:
Added (and also did the same for implicit_different_dims seeing as the file was still open).

arporter reviewed (changes requested):
Looking good - only fairly minor things now. A lot of the new test files set BASE_PATH and TEST_CONFIG but I think only one of them actually uses these?

One thing to consider: you've added/moved lots of test f90 files but our technology has now got to the point where it's probably easier and quicker to just generate the PSyIR from an in-line Fortran fragment in the test. I say this because the majority of those test files largely consist of the copyright/license with only ~10 lines of actual Fortran. However, I won't insist on this change as I don't want to prolong things any more than necessary.

All tests and examples are fine with compilation.
I'll check for conformance next time.

rupertford commented on src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:
```diff
@@ -0,0 +1,192 @@
+
+# Constants
+API = "nemo"
+BASE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
```
Well spotted. API is not used either.

rupertford commented on src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:
```diff
@@ -0,0 +1,192 @@
+    '''
+    assert NemoLoop()
+    nemo_loop = NemoLoop()
+    assert isinstance(nemo_loop, NemoLoop)
```
I've removed line 62

rupertford commented on src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:
```diff
@@ -0,0 +1,192 @@
+
+
+def test_create():
+    '''Test that the create static method creates a NemoLoop instance and
```
Updated this and one other occurance.

rupertford commented on src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:
```diff
@@ -0,0 +1,192 @@
+    assert children[0].parent is nemo_loop.loop_body
+    assert nemo_loop.loop_body.parent is nemo_loop
+    assert nemo_loop.variable is variable
+    assert nemo_loop.children[0] is start
```
This is done on purpose to check that the implementation is what we expect. However, you are right that I forgot to check `start`, `stop` and `step` so I've added these as well. If you think I'm being over the top then I'll remove the `nemo_loop.children*` lines.

arporter replied:
Yes, I think we should just check that the class behaves as expected. Testing how it's implemented seems to me to be overkill.

rupertford replied:
OK. Done.

rupertford commented on src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:
```diff
@@ -0,0 +1,192 @@
+
+
+def test_create_unknown():
+    '''Test that the create static method creates a NemoLoop instance with
```
Updated

rupertford commented on src/psyclone/tests/domain/nemo/nodes/nemo_loop_test.py:
```diff
@@ -0,0 +1,192 @@
+
+
+def test_create_errors():
+    '''Test that the expected exceptions are raises when the arguments to
```
Oops.

rupertford commented on src/psyclone/tests/domain/nemo/transformations/nemo_allarrayrange2loop_trans_test.py:
```diff
@@ -0,0 +1,168 @@
+    NemoArrayRange2LoopTrans and that it is a Transformation.
+
+    '''
+    assert NemoAllArrayRange2LoopTrans()
```
Agreed.

rupertford commented on src/psyclone/tests/domain/nemo/transformations/nemo_allarrayrange2loop_trans_test.py:
```diff
@@ -0,0 +1,168 @@
+    _, invoke_info = get_invoke("array_syntax.f90", api=API, idx=0)
+    schedule = invoke_info.schedule
+    for assignment in schedule.walk(Assignment):
+        trans = NemoAllArrayRange2LoopTrans()
```
Oh yes, well spotted.

rupertford commented on src/psyclone/tests/domain/nemo/transformations/nemo_allarrayrange2loop_trans_test.py:
```diff
@@ -0,0 +1,168 @@
+        "    tmask(ji,jj)=jpi\n"
+        "  enddo\n"
+        "enddo\n")
+    assert expected in result
```
Getting there gradually!

rupertford commented on src/psyclone/tests/domain/nemo/transformations/nemo_arrayrange2loop_trans_test.py:
```diff
@@ -0,0 +1,460 @@
+
+def test_apply_var_name():
+    '''Check that the variable name that is used when no names are
+    specified in the metadata does not clash with an existing symbol.
```
Yes, good point.

rupertford commented on src/psyclone/tests/domain/nemo/transformations/nemo_arrayrange2loop_trans_test.py:
```diff
@@ -0,0 +1,460 @@
+
+
+def test_not_outermost_range():
+    '''Check that the validate() method raises the expected exception if
```
It certainly isn't the first time and won't be the last :-)

rupertford commented on src/psyclone/tests/domain/nemo/transformations/nemo_outerarrayrange2loop_trans_test.py:
```diff
@@ -0,0 +1,180 @@
+    with pytest.raises(TransformationError) as info:
+        trans.validate(assignment)
+    assert("Transformation Error: Error in NemoOuterArrayRange2LoopTrans "
+           "transformation. The supplied assignment node should have an "
```
Actually you found a bug! The test was correct, it was just reporting the wrong node in the error message. I've now fixed the error message.

rupertford commented on src/psyclone/tests/domain/nemo/transformations/nemo_arrayrange2loop_trans_test.py:
```diff
@@ -0,0 +1,460 @@
+
+def test_apply_existing_names():
+    '''Check that the apply method uses existing iterators appropriately
+    when their symbols are already defined.
```
No it didn't. I've added code to check and associated tests.

rupertford commented:
Hopefully addressed the reviewers comments. Ready for next review from @arporter.

rupertford commented:
Taking this back as I forgot to remove any spurious BASI_PATH and TEST_CONFIG constants.

rupertford commented:
OK, ready for review by @arporter

arporter reviewed (changes requested):
Just two things left to do. I've brought the branch up-to-date with master and checked the compilation tests and examples.
The UG and Ref Guide also build fine.
Main changes are pylint/pycodestyle clean.

rupertford commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -330,6 +331,35 @@ def validate(self, node, options=None):
+                symbol_table = node.scope.symbol_table
+                loop_variable_symbol = symbol_table.lookup(loop_variable_name)
+                # Check the existing loop variable name is a scalar integer
+                if not (loop_variable_symbol.is_scalar and
```
That is possible and I modified the `validate` method to allow this even though I'm not sure that we want to support such a case. However, if so, when the transformation creates the explicit loop using `NemoLoop.create(...)`, this raises an exception as `create` expects the loop iterator to be a scalar integer. Therefore it is not possible to support this case, so instead I've modified the exception slightly to point out it might be a deferred type. Please let me know if you think there is a better option.

rupertford commented:
I've hopefully addressed the reviewer's comments and have brought the branch up-to-date with master. CI willing this is ready for the next review from @arporter.

arporter commented on src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py:
```diff
@@ -351,8 +351,8 @@ def validate(self, node, options=None):
-                        "the code as something that is not a scalar integer."
-                        "".format(loop_variable_name))
+                        "the code as something that is not a scalar integer, "
+                        "or is a deferred type.".format(loop_variable_name))
```
Sorry to be a pain but please could you add a test for this specific case - I suspect that you will need to check on the `type` of `loop_variable_symbol` since only a `DataSymbol` has the `is_scalar` method.

rupertford replied:
No, it's a fair comment. Actually DeferredType returns False if you call it with is_scalar, but that is not very nice behaviour to rely on so I've updated the if test as well as suggested.

arporter reviewed (changes requested):
Almost, almost there. I think that the case where a symbol is brought into scope via a USE needs a test and possible a code change.

rupertford commented:
Ready for next review subject to GHA completing happily

rupertford commented:
Back to @arporter

arporter reviewed (approved):
All requested changes have been made.
Tests and examples all OK with compilation.
UG and RG build cleanly.
I'll take the liberty of fixing a pylint error in one of the test files. (Unused os module imported.)
